contracts/optimize/SlotHelper.sol

```diff
@@ -0,0 +1,159 @@
+        mdata =  mdata << RIGHTSLOTDATASHIFT;
+        data = new bytes(len);
+        assembly{
+            mstore(add(data,0x20),mdata)
+        }
+    }
+
+    function getRaw(bytes32 key,bytes32 mdata) internal view returns(bytes memory res){
+        uint datalen ;
+        (datalen, res)= decodeMetadata1(mdata);
+        if (datalen > FIRSTSLOTDATASIZE){
+            uint ptr = 0;
+            assembly{
+                ptr := add(res,0x40) 
+            }
+
+            for (uint index = 0 ; index * 32 < datalen - FIRSTSLOTDATASIZE ; index ++){
+                assembly{
+                    
+                    mstore(0,add(key,index))
+                    let slot := keccak256(0,0x20)
+                    let cdata := sload(slot)
+                    
+                    // Or the last 4 bytes of the current word with the first 28 bytes of the previous word
+                    let value1 := shr(LENOFFSET,cdata)
+                    value1 := or(mload(sub(ptr,0x20)),value1)
+                    mstore(sub(ptr,0x20),value1)
+                    
+                    // Move the last 28 bytes of the current word forward by 32 bits
+                    let value2 := shl(RIGHTSLOTDATASHIFT , cdata)
+                    mstore(ptr,value2)
+
+                    ptr := add(ptr,0x20)
+                }
+            }
+        }
+    } 
+
+    function getRawAt(bytes32 key,bytes32 mdata, uint256 memoryPtr)
+        internal
+        view
+        returns (uint256 datalen, bool found)
+    {
+        bytes32 datapart;
+        (datalen, datapart)= decodeMetadata(mdata);
+
+        uint ptr = memoryPtr;
+        assembly{
+            mstore(ptr,datapart)
+            ptr := add(ptr,0x20) 
+        }
+
+        if (datalen > FIRSTSLOTDATASIZE){
+
+            for (uint index = 0 ; index * 32 < datalen - FIRSTSLOTDATASIZE ; index ++){
+                assembly{
+                    mstore(0,add(key,index))
+                    let slot := keccak256(0,0x20)
+                    let cdata := sload(slot)
+                    
+                    // Or the last 4 bytes of the current word with the first 28 bytes of the previous word
+                    let value1 := shr(LENOFFSET,cdata)
+                    value1 := or(mload(sub(ptr,0x20)),value1)
+                    mstore(sub(ptr,0x20),value1)
+                    
+                    // Move the last 28 bytes of the current word forward by 32 bits
+                    let value2 := shl(RIGHTSLOTDATASHIFT , cdata)
+                    mstore(ptr,value2)
+
+                    ptr := add(ptr,0x20)
+                }
+            }
+        }
+
+        found = true;
+    }
+
+    function isInSlot(bytes32 mdata) internal pure returns(bool succeed){
+        uint exist = uint256(mdata) >> LENOFFSET;
+        return exist > 0;
+    }
+
+    function encodeLen( uint datalen ) internal pure returns(bytes32 res){
+        assembly{
+            res := shl(LENOFFSET,datalen)
```

Comment:
Use << or >> instead.

Reply:
My consideration for using ASM here is that no additional type conversion is required

Reply:
I am not sure what is the meaning of type conversion here?  Should the code become

return datalen << LEN_OFFSET?

Reply:
extra type conversion: uint=> bytes32
```
 function encodeLen( uint datalen ) internal pure returns(bytes32 res){
        res = bytes32(datalen  << LENOFFSET); 
    }
```

Reply:
https://github.com/QuarkChain/evm-large-storage/pull/1/commits/6ead79e1295b80e588004fc5e719f0b7e6e41bb9